lib/THC/THCTensorMath2.cu

```diff
@@ -660,3 +660,55 @@ void THCudaTensor_randn(THCState *state, THCudaTensor *r_, THLongStorage *size)
   THCudaTensor_resize(state, r_, size, NULL);
   THCudaTensor_normal(state, r_, 0, 1);
 }
+
+__global__ void THCudaTensor_kernel_cross(float *src_, float *x, float *y, long nelem) {
+  long blockid = blockIdx.x + blockIdx.y * gridDim.x + gridDim.x*gridDim.y*blockIdx.z;
+  long start = 3*(blockid*blockDim.x + threadIdx.x);
+
+  long p1 = start + 0;
+  long p2 = start + 1;
+  long p3 = start + 2;
+  if (p3 > nelem) return;
+  src_[p1] = x[p2]*y[p3] - x[p3]*y[p2];
+  src_[p2] = x[p3]*y[p1] - x[p1]*y[p3];
+  src_[p3] = x[p1]*y[p2] - x[p2]*y[p1];
+}
+
+THC_API void THCudaTensor_cross(THCState *state, THCudaTensor *self, THCudaTensor *x, THCudaTensor *y, int dimension)
+{
+  THAssert(THCudaTensor_checkGPU(state, 3, self, x, y));
+
+  int i;
+  long nd = THCudaTensor_nDimension(state, x);
+  long nelem = THCudaTensor_nElement(state, x);
+  THArgCheck(nd == THCudaTensor_nDimension(state, y), 1, "tensors must have same number of dimensions");
+  for (i=0; i<nd; i++) {
+    THArgCheck(THCudaTensor_size(state, x, i) == THCudaTensor_size(state, y, i), 1, "dimension %i of x and y does not match", i);
+    if (dimension < 0 && THCudaTensor_size(state, x, i) == 3) dimension = i; 
+  }
+
+  THArgCheck(dimension >= 0 && dimension < nd, 3, "dimension %d out of range", dimension+1);
+  THArgCheck(THCudaTensor_size(state, x, dimension) == 3, 3,
+      "dimension %d does not have size 3", dimension+1);
+
+  THCudaTensor_transpose(state, x, NULL, dimension, nd-1);
+  THCudaTensor_transpose(state, y, NULL, dimension, nd-1);
+  THCudaTensor_resizeAs(state, self, x);
+  THCudaTensor *nx = THCudaTensor_newContiguous(state, x);
+  THCudaTensor *ny = THCudaTensor_newContiguous(state, y);
+  THCudaTensor_transpose(state, x, NULL, dimension, nd-1);
+  THCudaTensor_transpose(state, y, NULL, dimension, nd-1);
+
+  long threadsize = 192;
+  long MG = 65535;
+  dim3 threads(threadsize);
+  dim3 grid(min(MG, nelem), min(MG, nelem/MG + 1), min(MG, nelem/MG/MG));
```

Comment:
The grid is not taking into account `threadsize` at all; the kernel is over-launching threads. 
If the input tensor `x` has 12 elements, you will launch a grid with 192 threads x 4 blocks, whereas a single block will suffice.

Also, a construction like `nelem/MG +1` is overlaunching as well if nelem is an even multiple of MG. This is what the divup construction is for.

Fixing the number of threads you launch is ok, but the number of blocks that you launch should be related to `divup(nelem, (threadsize * 3))`.

Instead of doing this kind of math, it might make more sense to have the kernel itself loop in a grid-stride loop

https://devblogs.nvidia.com/parallelforall/cuda-pro-tip-write-flexible-kernels-grid-stride-loops/

Even better, I think you can actually use the THCTensorApply templates, you just need to transpose the size-3 dimension to innermost (which you are doing), but then set the size of that dimension to 1 in the TensorInfo; that will allow you to stride by 3. You could either ensure that the size-3 dimension is also contiguous and access the other elements by the points you are given, `p + 1` and `p + 2`, or use the stride for that dimension to get the other points.

Of course, the kernel will be fastest if the size-3 dimension is contiguous in memory. The transposes might still be desired, but the pattern with the rest of cutorch is that we use the THCTensorApply stuff to operate on non-contiguous tensors directly without newContiguous, to avoid cudaMalloc.